### DISCUSSION TITLE
Abnormal value for "Battery Total charge"

### DISCUSSION THREAD

**Discussion by JensenNick** [wlcrs/huawei_solar#628] [Setting up the integration]
After a somewhat intense 6 month period developing the "Huawei Solar - PEES" (energy and economy sensors I've shared [here](https://github.com/JensenNick/huawei_solar_pees)), the value for my "Battery Total charge" / `sensor.battery_total_charge` has reached the abnormal value of 461 thousand kWh. 

Not sure why, how, or when this has occurred, but I've admittedly have had to reinstall the integration several times due to errors on my part in this period and I've had sensors "jump". I've contacted my installer who say that it looks all right at their end and according to the FusionSolar app the value is be approx. 1.800 kWh (I can't find the energy charge in the web interface - only discharge).

I've tried to uninstall the integration, delete statistics and purge entities related to the integration, reboot and reinstall the integration - but the value stays at this abnormal value.

What would be the best way to go about getting this value corrected for a consistency in HA? Would it be sufficient to "just" correct the database?

@wlcrs Not sure if this has something to do with the code or not. Let me know if you'd like me to do some test / report.

![2024-02-24_08-21-49](https://github.com/wlcrs/huawei_solar/assets/149193835/3acd01e2-ec96-49a9-bc88-1c1b6e370af3)

**Comment by JensenNick** [comment]
This issue persists. I believe I've tried everything. After deleting the integration I've even deleted the database (state, statistics and statistics_short_term), purged the sensors `battery_total_charge` and `battery_total_discharge` made sure they are not in STATES nor in STATISTICS, rebooted HA - but once the integration is reinstalled it starts of at this abnormal level every time. I'd really appreciate some help to fix this. @wlcrs ?
![2024-04-15_13-34-57](https://github.com/wlcrs/huawei_solar/assets/149193835/669e772d-f802-46fd-9ba1-1ed7e8671bd5)

**Comment by torwag** [comment]
I have the same issue. My system just went online and for a few seconds I was fearing that they sold me some used battery blocks ;)
Currently, I get    WR1-Battery Total charge 503,048.04 kWh.
Interestingly and oddly, I only get this for the first battery connected to the master inverter, whereas the other two batteries connected to the slave inverters seems to have correct values of about 300kWh. That tells me that the reading itself, seems to be not just a conversion error on the add-on side, as I assume that those values get read-in in the same way.

**Reply by torwag** [reply]
Is there a way like e.g. a debug logger option to get current readings in raw format, thus we can check what actually goes wrong?

**Reply by Roving-Ronin** [reply]
@torwag you can install https://sourceforge.net/projects/qmodmaster/ and connect to the Inverter+Battery etc, directly using modbus.
Remember to disable the WLCRS integration, given the sDongle only supports a single modbus connection at a time.

Settings to use for battery #1 'Total Discharge" are:

![image](https://github.com/wlcrs/huawei_solar/assets/108674933/b3550a06-4bef-44e4-8576-a9965d207952)

**Reply by torwag** [reply]
@Roving-Ronin thanks for the feedback. Unfortunately, qmodmaster seems to have trouble to run under linux/sway/wayland. I can try some python libs and create a few line of code to read out the inverters.
Just one question regarding the Sdongle. You said, I have to disable the WLCRS integration as the SDongle can only support one connection. I connect via the three wifi AP of the inverters. Would that cause the same issue?

**Reply by Roving-Ronin** [reply]
@torwag it rings a bell that the inbuilt AP also only support one connection at a time.

**Comment by JensenNick** [comment]
@wlcrs I'm trying to resolve this issue but I can not do it without your help. I've tried to reinstall the integration from scratch several times also on completely fresh installs of HA. The initial value is still abnormal.

I'm now in contact with Huawei Support who insist that they need to look at how the integration get / interpret the modbus data. They are asking for "the capture package", "data frame package" or "the binary data" sent between the integration and the Huawei dongle. I had a meeting with them where they drew this diagram to explain.
![2024-11-15_10-46-45](https://github.com/user-attachments/assets/a64b59af-f3bb-49b7-87e0-594cab23abd1)

Here is what they need in their own words
![2024-11-19_16-36-23](https://github.com/user-attachments/assets/27bbccff-9e4c-4526-b740-76dabe720ffe)

I have no clue where and how to obtain the "capture package" that they mention in bullet #1. They seem to be very convinced that the first thing they need to do is to make sure that the reading of data to the integration corresponds to the data read to FusionSolar. Please help.

**Reply by torwag** [reply]
What they mean is that the packages send from the inverter to the third party app (wlcrs plugin) should be logged. Package means really the raw data before it gets parsed and formated in any way.
That could be done by sniffing the communication between inverter and wlcr, or (depending of how that data is obtained) by adding a logging function in that library which really does the bottom work of reading in data packages.

What is odd, for me only the master inverter (that with the sd-dongle) provide wrong data for the battery. However the batteries of the other two inverters is fine. I assume that the entire data pathway is identical in the way data is read out and converted. Thus, it must be that wrong data is send.

And I agree for this spoofing the data packages would be very useful.

**Reply by JensenNick** [reply]
@wlcrs Could you please help me a bit here... Thanks!
![image](https://github.com/user-attachments/assets/a508b474-9f7a-4e41-8e4f-019cb91bb396)

**Comment by wlcrs** [comment]
Please create a bug report and properly fill it out. As per the instructions in the form, attach a debugging log that contains a few minutes of data, allowing me to see a few requests of that modbus register.

**Comment by wlcrs** [comment]
Closing this in favor of #856
